### PR TITLE
Testcase: Snapshot creation and restore of PGSQL Workloads PVCs

### PR DESCRIPTION
**Pull request by Akarsha-rai**
Signed-off-by: Akarsha <akrai@redhat.com>

### CODE REVIEW

File: tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                f"Created PVC {restore_pvc_obj.name} from snapshot "
+                f"{snapshot.name}"
+            )
+            restore_pvc_objs.append(restore_pvc_obj)
+        log.info("Created new PVCs from all the snapshots and in Bound state")
+
+        # Attach a new pgsql pod to created pvc's and run pgbench benchmark on pod
+        i = 0
+        pgsql_obj_list = []
+        for pvc_obj in restore_pvc_objs:
```
Comment (PrasadDesala) [inline tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py:79]: may be we can get the length of restore_pvc_objs and then iterate over its length.
This way you dont need to initialize i and increment it inside the function
WDYT?

File: tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            try:
+                pgsql_sset = templating.load_yaml(
+                    constants.PGSQL_STATEFULSET_YAML
+                )
+                del pgsql_sset['spec']['volumeClaimTemplates']
+                pgsql_sset['metadata']['name'] = 'postgres-snap' + f"{i}"
+                pgsql_sset['spec']['template']['spec']['containers'][0]['volumeMounts'][0]['name'] = pvc_obj.name
+                pgsql_sset['spec']['template']['spec']['volumes'] = [
+                    {'name': f'{pvc_obj.name}', 'persistentVolumeClaim': {'claimName': f'{pvc_obj.name}'}}
```
Comment (PrasadDesala) [inline tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py:88]: Please break the lines

File: tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                time.sleep(wait_time)
+
+            except (CommandFailed, CalledProcessError) as cf:
+                log.error('Failed during creation of postgres pod')
+                raise cf
+
+        pgsql.wait_for_pgbench_status(status=constants.STATUS_COMPLETED, timeout=1800)
+
+        # Delete created postgres and pgbench pods
+        log.info("Deleting created postgres pods and pgbench configuration")
```
Comment (PrasadDesala) [inline tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py:117]: You are deleting only the newly created postgres pods, pgbench clean up is taken care by pgsql.cleanup()
Please modify log info message accordingly

File: tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py (hunk 2 above)
Comment (PrasadDesala) [inline tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py:88]: Can we create a function of this? we might follow the same steps even for other snapshot related tests.

File: tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+        # Create PVCs out of the snapshots
+        log.info("Creating new PVCs from snapshots")
+        restore_pvc_objs = []
+        for snapshot in snapshots:
+            log.info(f"Creating a PVC from snapshot {snapshot.name}")
+            restore_pvc_obj = snapshot_restore_factory(
+                snapshot_obj=snapshot,
+                restore_pvc_name=f"{snapshot.name}-restored",
+                size='10Gi',
```
Comment (Avilir) [inline tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py:63]: it is better to take this value from the created PVC (original) and not use a constant.

File: ocs_ci/templates/workloads/pgsql/server/StatefulSet.yaml [hunk 5]
```diff
@@ -34,6 +34,7 @@ spec:
           volume.beta.kubernetes.io/storage-class: ocs-storagecluster-ceph-rbd
```
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply ocs_ci/templates/workloads/pgsql/server/StatefulSet.yaml]: removed

File: tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    skipif_ocs_version, E2ETest, tier1
+)
+from ocs_ci.ocs.resources.ocs import OCS
+from ocs_ci.utility import templating
+from ocs_ci.ocs.exceptions import CommandFailed
+
+log = logging.getLogger(__name__)
+
+
+@tier1
```
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py]: changed to tier2

File: tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py (hunk 1 above)
Comment (Akarsha-rai) [reply tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py]: moved to pgsql file

File: tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py (hunk 2 above)
Comment (Akarsha-rai) [reply tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py]: moved to pgsql file

File: tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py (hunk 4 above)
Comment (Akarsha-rai) [reply tests/e2e/lifecycletest/pvc_snap_clone/test_workload_pvc_snapshot.py]: agreed, removed and directly gets from Parent PVC

**Review by Avilir — DISMISSED** [review]
LGTM

File: ocs_ci/ocs/pgsql.py [hunk 7]
```diff
@@ -461,3 +474,64 @@ def cleanup(self):
...
+
+        """
+        pgsql_obj_list = []
+        for pvc_obj in pvc_objs:
+            try:
+                pgsql_sset = templating.load_yaml(
+                    constants.PGSQL_STATEFULSET_YAML
+                )
+                del pgsql_sset['spec']['volumeClaimTemplates']
+                pgsql_sset['metadata']['name'] = f"{postgres_name}" + f"{pvc_objs.index(pvc_obj)}"
```
Comment (PrasadDesala) [inline ocs_ci/ocs/pgsql.py:501]: You are pass the postgres arg name as None, what if someone doesn't pass the postgres name to the function?
The name according to here will be just index.
Please check.

File: ocs_ci/ocs/pgsql.py [hunk 8]
```diff
@@ -461,3 +474,64 @@ def cleanup(self):
...
+                ]
+                pgsql_sset = OCS(**pgsql_sset)
+                pgsql_sset.create()
+                pgsql_obj_list.append(pgsql_sset)
+
+                self.wait_for_postgres_status(status=constants.STATUS_RUNNING, timeout=300)
+
+                if run_benchmark:
+                    pg_data = templating.load_yaml(constants.PGSQL_BENCHMARK_YAML)
+                    pg_data['metadata']['name'] = f"{pgbench_name}" + f"{pvc_objs.index(pvc_obj)}"
```
Comment (PrasadDesala) [inline ocs_ci/ocs/pgsql.py:517]: same as above comment

File: ocs_ci/ocs/pgsql.py (hunk 7 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/pgsql.py:501]: done, added as mandatory to send the postgres_name

File: ocs_ci/ocs/pgsql.py (hunk 8 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/pgsql.py]: thank you for pointing out, changed now

**Review by jilju — APPROVED** [review]

**Review by PrasadDesala — APPROVED** [review]

**Pull request merged**